Add the migration guide for shadow library (#1226)
aws/aws-iot-device-sdk-embedded-C: docs/doxygen/migration/shadow.txt

/**
@page shadow_migration Shadow
@brief How to migrate a Shadow application from v3 to 202009.00.

> The Device Shadow library has been redesigned in this release.  The new design has the following differences than the v3 of the AWS IoT Device SDK for Embedded C.

Reviewer: ```suggestion
The Device Shadow library has been redesigned in this release.  The new design has the following differences as compared to the Device Shadow library in v3 of the AWS IoT Device SDK for Embedded C.
```

Author: plz check [the commit](https://github.com/aws/aws-iot-device-sdk-embedded-C/pull/1226/commits/f504c65a1d8030ae4237312b7b63b563b8700ca8).

> - The new Device Shadow library is completely decoupled from an underlying MQTT library. The v3 library depends on the same version of the MQTT library in the AWS IoT Device SDK for Embedded C. <br>

Reviewer: ```suggestion
- The new Device Shadow library is completely decoupled from the underlying MQTT Client library. The v3 Device Shadow library, on the other hand, depended on the MQTT Client library present in the v3 of AWS IoT Device SDK for Embedded C. <br>
```

Author: plz check [the commit](https://github.com/aws/aws-iot-device-sdk-embedded-C/pull/1226/commits/f504c65a1d8030ae4237312b7b63b563b8700ca8).

> - The new library does not have any code for handling JSON files. The v3 library contained code for assembling simple JSON files in the format expected by the Device Shadow service. <br>

Reviewer: ```suggestion
- The new Device Shadow library does not have any code for constructing JSON files. The v3 Device Shadow library, on the other hand, contained code for constructing JSON files in the format expected by the AWS IoT Device Shadow Service. <br>
```

Author: plz check [the commit](https://github.com/aws/aws-iot-device-sdk-embedded-C/pull/1226/commits/f504c65a1d8030ae4237312b7b63b563b8700ca8).

> - The new library only provides simple convenience macros and functions for handling MQTT topic strings defined by AWS IoT Device Shadow service.  It does not contain any functionality of a MQTT stack. The v3 library, on the other hand, provided a complete set of APIs for interacting with the service by using a MQTT stack underneath.<br>

Reviewer: ```suggestion
- The new Device Shadow library provides simple convenience macros and functions for handling MQTT topic strings as defined by AWS IoT Device Shadow Service. The v3 Device Shadow library, on the other hand, provided a complete set of APIs for interacting with the AWS IoT Device Shadow Service by using a MQTT Client library.<br>
```

Author: plz check [the commit](https://github.com/aws/aws-iot-device-sdk-embedded-C/pull/1226/commits/f504c65a1d8030ae4237312b7b63b563b8700ca8).

> Regardless of the choice of the MQTT library, the application can use the new Device Shadow library by following the call sequence diagram in the [Design document](@ref shadow_design).

Reviewer: ```suggestion
The application can use the new Device Shadow library with their choice of MMQTT Client library by following the call sequence diagram in the [Design document](@ref shadow_design).
```

Author: plz check [the commit](https://github.com/aws/aws-iot-device-sdk-embedded-C/pull/1226/commits/f504c65a1d8030ae4237312b7b63b563b8700ca8).

Author: plz check [the commit](https://github.com/aws/aws-iot-device-sdk-embedded-C/pull/1226/commits/f504c65a1d8030ae4237312b7b63b563b8700ca8).

/**
@page shadow_migration Shadow

> @brief How to migrate an Shadow application from v3 to 202009.00.

Author: will fix it.

> @brief How to migrate an Shadow application from v3 to 202009.00.

Author: plz check [the commit](https://github.com/aws/aws-iot-device-sdk-embedded-C/pull/1226/commits/0f6b5f0f9bdb5f6fd2ce1e55545e99e6c3563f2d).

> The Device Shadow library has been redesigned in the Long Term Support (LTS) release.  The new design has the following differences than the Version 3 of the AWS IoT Device SDK for Embedded C.

Author: How about change to:
The Device Shadow library has been redesigned in `this` release.

- The new Device Shadow library is completely decoupled from an underlying MQTT library.  The Version 3 library depended on the same version of MQTT library in the AWS IoT Device SDK for Embedded C. <br>
- The new library does not have any code for handling JSON files.  The Version 3 library contained code for assembling simple JSON files in the format expected by the Device Shadow service. <br>
- The new library only provides simple convenience macros and functions for handling MQTT topic strings defined by AWS IoT Device Shadow service.  It does not contain any functionality of a MQTT stack.  The Version 3 library, on the other hand, provided a complete set of API for interacting with the service by using a MQTT stack underneath. <br>

An application can continue using the Version 3 Device Shadow library if it also chooses to keep using the same version of MQTT library of the AWS IoT Device SDK for Embedded C.

Regardless an application uses what MQTT library of its choice, the application can use the new Device Shadow library by following the call sequence diagram in the

> [Design document](@ref shadow_design).

Reviewer 2: "Regardless of the choice of the MQTT library, the application can use ... "

/**
@page shadow_migration Shadow
@brief How to migrate an Shadow application from v3 to 202009.00.

The Device Shadow library has been redesigned in the Long Term Support (LTS) release.  The new design has the following differences than the Version 3 of the AWS IoT Device SDK for Embedded C.

> - The new Device Shadow library is completely decoupled from an underlying MQTT library.  The Version 3 library depended on the same version of MQTT library in the AWS IoT Device SDK for Embedded C. <br>

Reviewer 3: ```suggestion
- The new Device Shadow library is completely decoupled from an underlying MQTT library.  The Version 3 library depends on the same version of the MQTT library in the AWS IoT Device SDK for Embedded C. <br>
```

Reviewer 3: Also can you change all Version 3 to "v3"?

Author: I also think the first sentence is enough. will remove the one about v3.

- The new library does not have any code for handling JSON files.  The Version 3 library contained code for assembling simple JSON files in the format expected by the Device Shadow service. <br>

> - The new library only provides simple convenience macros and functions for handling MQTT topic strings defined by AWS IoT Device Shadow service.  It does not contain any functionality of a MQTT stack.  The Version 3 library, on the other hand, provided a complete set of API for interacting with the service by using a MQTT stack underneath. <br>

Reviewer 3: ```
- The new library only provides simple convenience macros and functions for handling MQTT topic strings defined by AWS IoT Device Shadow service.  It does not contain any functionality of a MQTT stack.  The Version 3 library, on the other hand, provided a complete set of APIs for interacting with the service by using a MQTT stack underneath. <br>
```

Reviewer 2: Overall comment on the doc - do we need to contrast and compare this to v3?

Reviewer 4: This is the migration guide for users of v3 API.  That's why there is a description of the differences.  In the MQTT's migration guide, the differences are presented in a table of function-to-function comparison.  The Shadow is completely different - there is no function-to-function mapping at all because of the architecture change, therefore the differences are presented as a narrative.

Reviewer 2: Ok thats fine then. It felt a bit odd, but we can polish it once its merged.

Author: plz check [the commit](https://github.com/aws/aws-iot-device-sdk-embedded-C/pull/1226/commits/0f6b5f0f9bdb5f6fd2ce1e55545e99e6c3563f2d) for the update.

> An application can continue using the Version 3 Device Shadow library if it also chooses to keep using the same version of MQTT library of the AWS IoT Device SDK for Embedded C.

Reviewer 3: ```suggestion
An application can continue using the Version 3 Device Shadow library only if it also keeps using the same version of the MQTT library in the AWS IoT Device SDK for Embedded C.
```

Author: plz check [the commit](https://github.com/aws/aws-iot-device-sdk-embedded-C/pull/1226/commits/0f6b5f0f9bdb5f6fd2ce1e55545e99e6c3563f2d).

> Regardless an application uses the Version 3 MQTT library or any MQTT library of its choice, the application can use the new Device Shadow library by following the call sequence diagram in the [Design document](@ref shadow_design).

Reviewer 3: ```suggestion
An application may use the version 3 MQTT library, version 202009.00 MQTT library, or any other MQTT library with the new Device Shadow library. To use the new Device Shadow library with any MQTT library, please see the call sequence diagram in [Design document](@ref shadow_design). 
```

Author: plz check [the commit](https://github.com/aws/aws-iot-device-sdk-embedded-C/pull/1226/commits/0f6b5f0f9bdb5f6fd2ce1e55545e99e6c3563f2d).

@section migration_shadow_removed Removed features
@brief The following features were present in v3, but removed in 202009.00:
- <u>MQTT regarding functions</u> <br>
  Version 3 has a series of functions which are on top of the MQTT. Considering to make Shadow as an independent library, those MQTT regarding functions won't be used anymore.
- <u>JSON regarding functions</u> <br>
  Version 3 has a series of functions which are on top of the JSON. Considering to make Shadow as an independent library, those JSON regarding functions won't be used anymore.
- <u>callback function</u> <br>
  Version 3 has a series of callback functions to receive the state changes on the regarding shadow topics. In version 4, this series of callback functions has been removed. <br>
  <b>New Design:</b> In version 4, `Shadow_MatchTopic` will be used to determine whether it is related to a device shadow; if it is, return information about the type of device shadow message.
- <u>Thread support</u> <br>
  Version 3 has an `_ENABLE_THREAD_SUPPORT_` macro for multi-threading use cases. Version 4 has been redesigned to use only a single execution thread, and is
  not thread-safe.

> @section migration_shadow_datatypes Data Types

Reviewer 3: If there are no equivalent data types please delete this section

Author: remove in [this commit](https://github.com/aws/aws-iot-device-sdk-embedded-C/pull/1226/commits/19dedddc029d17d1c5428d938981b1216acd1452).

@brief The following table lists equivalent data types in v3 and 202009.00.

<table>
    <tr>
        <th>Version 3</th>
        <th>Version 4</th>
        <th>Notes</th>
    </tr>
    <tr>
        <td>None</td>
        <td>None</td>
        <td>None</td>
    </tr>
</table>

> @section migration_shadow_functions Functions

Reviewer 3: If there are no equivalent functions please delete this section.

Author: remove in [this commit](https://github.com/aws/aws-iot-device-sdk-embedded-C/pull/1226/commits/19dedddc029d17d1c5428d938981b1216acd1452).
